power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,255 @@
+    auto begin() const { return group_iterator(0); }
+    auto end() const { return group_iterator(size()); }
+    auto get_element_range(Idx group) const { return *group_iterator(group); }
+
+    auto element_size() const { return indptr_.back(); }
+    auto get_group(Idx element) const -> Idx {
+        assert(element < element_size());
+        return std::distance(std::begin(indptr_), std::upper_bound(std::begin(indptr_), std::end(indptr_), element)) -
+               1;
+    }
+
+    SparseGroupedIdxVector() = default;
+    explicit SparseGroupedIdxVector(IdxVector sparse_group_elements)
+        : indptr_{sparse_group_elements.empty() ? IdxVector{0} : std::move(sparse_group_elements)} {
+        assert(size() >= 0);
+        assert(element_size() >= 0);
+        assert(std::is_sorted(std::begin(indptr_), std::end(indptr_)));
+    }
+    SparseGroupedIdxVector(from_sparse_t /* tag */, IdxVector sparse_group_elements)
+        : SparseGroupedIdxVector{std::move(sparse_group_elements)} {}
+    SparseGroupedIdxVector(from_dense_t /* tag */, IdxVector const& dense_group_elements, Idx num_groups)
+        : SparseGroupedIdxVector{detail::sparse_encode(dense_group_elements, num_groups)} {}
+
+  private:
+    IdxVector indptr_;
+};
+
+class DenseGroupedIdxVector {
+  private:
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
+                                                        boost::iterator_range<IdxCount>, Idx> {
+      public:
+        using iterator = boost::iterator_range<IdxCount>;
+
+        explicit constexpr GroupIterator(IdxVector const& dense_vector, Idx const& group)
+            : dense_vector_{dense_vector},
+              group_{group},
+              group_range_{std::equal_range(std::cbegin(dense_vector_), std::cend(dense_vector_), group)} {}
+
+      private:
+        using group_iterator = IdxVector::const_iterator;
+
+        IdxVector const& dense_vector_;
+        Idx group_;
+        std::pair<group_iterator, group_iterator> group_range_;
+
+        friend class boost::iterator_core_access;
+
+        boost::iterator_range<IdxCount> dereference() const {
+            return boost::counting_range(
+                detail::narrow_cast<Idx>(std::distance(std::cbegin(dense_vector_), group_range_.first)),
+                detail::narrow_cast<Idx>(std::distance(std::cbegin(dense_vector_), group_range_.second)));
+        }
+        constexpr bool equal(GroupIterator const& other) const { return group_ == other.group_; }
+        constexpr void increment() { advance(1); }
+        constexpr void decrement() { advance(-1); }
+        constexpr void advance(Idx n) {
+            auto const start = n > 0 ? group_range_.second : std::cbegin(dense_vector_);
+            auto const stop = n > 0 ? std::cend(dense_vector_) : group_range_.first;
+
+            group_ += n;
+            group_range_ = std::equal_range(start, stop, group_);
```

Comment:
This is exactly we don't want to implement `advance` to satisfy random access iterator.

If we only implement `increment` to satistfy forware iterator. The logic here can be much simpler.